### PR TITLE
Feat/pandas reader parquet

### PR DESCRIPTION
**Pull request by flaviassantos**
This pull request adds parquet read and write functionality for issue  #406.

## Changes

The changes to the files are as described below:

`pandas_extensions.py`: Added the classes to read and write parquet files
`test_pandas_extensions.py`: Added a single test case that exercises the writing and reading functionality respectively
`notebook.ipynb`: Added the example of parquet materialization
`my_script.py`: Added the example of parquet materialization

## How I tested this

By running successfully:
- unit test for the PandasParquetWriter and PandasParquetReader classes.
- the Jupyter notebook.
- the my_script.py
- cicleci job locally

## Checklist

- [x] PR has an informative and human-readable title (this will be pulled into the release notes)
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future TODOs are captured in comments
- [ ] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

File: hamilton/plugins/pandas_extensions.py [hunk 1]
```diff
@@ -126,20 +126,91 @@ def name(cls) -> str:
...
+    compression: Optional[str] = "snappy"
+    index: Optional[bool] = None
+    partition_cols: Optional[List[str]] = None
+    storage_options: Optional[Dict[str, Any]] = None
+
+    @classmethod
+    def applicable_types(cls) -> Collection[Type]:
+        return [DATAFRAME_TYPE]
+
+    def _get_saving_kwargs(self) -> Dict[str, Any]:
```
Comment (flaviassantos) [inline hamilton/plugins/pandas_extensions.py:197]: I need guidance on how to implement the **kwargs that are not listed in the Pandas [docs](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_parquet.html#pandas.DataFrame.to_parquet) nor in the [user guide](https://pandas.pydata.org/docs/user_guide/io.html#parquet)

**Comment by flaviassantos** [comment]
@skrawcz , last point in the checklist. Can you point me to the "Project documentation" mentioned?

File: hamilton/plugins/pandas_extensions.py (hunk 1 above)
Comment (skrawcz) [reply hamilton/plugins/pandas_extensions.py:197]: yep. So, options:

1. Determine what those kwargs are and surface them as fields in the class.
2. add a `extra_kwargs` dictionary to the class so people can explicitly pass them in, and then add it to the kwargs argument.

I think (2) is probably simpler.

```python
extra_kwargs: Optional[Dict[str, Any]] = None

...

if self.extra_kwargs is not None:
   kwargs.update(self.extra_kwargs)
```

**Comment by skrawcz** [comment]
> @skrawcz , last point in the checklist. Can you point me to the "Project documentation" mentioned?

Yep so that should be automatically updated for you. It should show up here (based on the build of this branch) https://hamilton--429.org.readthedocs.build/en/429/reference/io/available-data-adapters/.

File: tests/plugins/test_pandas_extensions.py [hunk 2]
```diff
@@ -30,6 +32,19 @@ def df():
...
+def test_pandas_parquet(df: pd.DataFrame, tmp_path: pathlib.Path) -> None:
+    file_path = tmp_path / "sample_df.parquet.gzip"
+    writer = PandasParquetWriter(path=file_path, engine="pyarrow")
+    writer.save_data(df)
+
+    reader = PandasParquetReader(path=file_path, engine="pyarrow")
+    read_df, metadata = reader.load_data(pd.DataFrame)
+
+    assert read_df.shape == df.shape, "DataFrames do not match"
+    assert read_df.columns == df.columns
```
Comment (skrawcz) [inline tests/plugins/test_pandas_extensions.py:44]: Pro-tip: you probably want to use https://pandas.pydata.org/docs/reference/api/pandas.testing.assert_frame_equal.html

**Review by skrawcz — COMMENTED** [review]
looks good, just the minor comment on the test!

**Comment by skrawcz** [comment]
> try:
>          result = self.api.parquet.read_table(
>               path_or_handle, columns=columns, **kwargs
>         ).to_pandas(**to_pandas_kwargs)
> E           TypeError: read_table() got an unexpected keyword argument 'dtype_backend'
>
>../venvs/hamilton-venv/lib/python3.7/site-packages/pandas/io/parquet.py:240: TypeError

You will need to gate this parameter for 3.7. So you'll see others have:

```python
        if sys.version_info >= (3, 8) and self.dtype_backend is not None:
            kwargs["dtype_backend"] = self.dtype_backend
```

File: hamilton/plugins/pandas_extensions.py (hunk 1 above)
Comment (flaviassantos) [reply hamilton/plugins/pandas_extensions.py:198]: Thanks, that is a simpler option. Although should we handle any errors when a user enters an unexpected keyword argument?

File: tests/plugins/test_pandas_extensions.py (hunk 2 above)
Comment (flaviassantos) [reply tests/plugins/test_pandas_extensions.py:44]: Thanks :), yeah I looked into this. When data is read from parquet, the string(python) in the original df became an object in the read_df...so I needed to pass a 'check_dtype=False' in other to pass the test. 

Is it a workaround or does it sounds right?

**Comment by flaviassantos** [comment]
> try: result = self.api.parquet.read_table( path_or_handle, columns=columns, **kwargs ).to_pandas(**to_pandas_kwargs) E TypeError: read_table() got an unexpected keyword argument 'dtype_backend'
> 
> ../venvs/hamilton-venv/lib/python3.7/site-packages/pandas/io/parquet.py:240: TypeError

@skrawcz, thanks! 
Added the conditional to the class. For future reference, how do you determined/run this test to see if it was backwards compatible?

File: hamilton/plugins/pandas_extensions.py (hunk 1 above)
Comment (skrawcz) [reply hamilton/plugins/pandas_extensions.py:198]: No I think that's on them unless it's something really common. We can always add this later.

File: tests/plugins/test_pandas_extensions.py (hunk 2 above)
Comment (skrawcz) [reply tests/plugins/test_pandas_extensions.py:46]: ah -- that's fine. Likely something to look into how to get them to match easily. But for this test we just want to check that it looks right.

**Review by skrawcz — APPROVED** [review]

**Pull request merged**